File: python/random_forest.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 25 12:44:59 2025

Random Forest Regression on Position Salaries Dataset
Author: Chukwuka Chijioke Jerry
"""

# ===========================
# 📦 Import Libraries
# ===========================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestRegressor

# ===========================
# 📂 Load Dataset
# ===========================
dataset = pd.read_csv('../data/Position_Salaries.csv')
dataset = dataset.iloc[:, 1:]  # Keep only Level and Salary columns
X = dataset.iloc[:, [0]].values
y = dataset.iloc[:, -1].values

# ===========================
# 🌳 Fit Random Forest Regression
# ===========================
regressor = RandomForestRegressor(
    n_estimators=100, random_state=0
)
regressor.fit(X, y)

# ===========================
# 🔍 Prediction
# ===========================
y_pred = regressor.predict([[6.5]])
print(f"💰 Predicted Salary for Level 6.5: {y_pred[0]:.2f}")

# ===========================
# 📊 Visualization
# ===========================
plt.scatter(X, y, color='red')
plt.plot(X, regressor.predict(X), color='blue')
plt.title('Random Forest Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.grid(True)
plt.tight_layout()
plt.show()

# ===========================
# 🖼️ High-Resolution Plot
# ===========================
X_grid = np.arange(X.min(), X.max(), 0.001).reshape(-1, 1)

plt.scatter(X, y, color='red')
plt.plot(X_grid, regressor.predict(X_grid), color='blue')
plt.title('Random Forest Regression (High Resolution)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.grid(True)
plt.tight_layout()
plt.show()

File: r/random_forest.R
# ============================================
# 📊 Random Forest Regression
# Author: Chukwuka Chijioke Jerry
# ============================================

# ===========================
# 📂 Load Dataset
# ===========================
dataset <- read.csv('../data/position_salaries.csv')
dataset <- dataset[2:3]  # Keep only Level and Salary columns

# ===========================
# 🌳 Fit Random Forest Model
# ===========================
library(randomForest)
set.seed(1234)

# Fit Random Forest with 500 trees
regressor <- randomForest(
  x = dataset[1],           # Level as feature
  y = dataset$Salary,       # Salary as target
  ntree = 500
)

# ===========================
# 🔍 Prediction
# ===========================
y_pred <- predict(
  regressor,
  newdata = data.frame(Level = 6.5)
)
cat("💰 Predicted Salary at level 6.5:", y_pred, "\n")

# ===========================
# 📈 Visualization (Training data)
# ===========================
library(ggplot2)

ggplot() +
  geom_point(
    aes(x = dataset$Level, y = dataset$Salary),
    color = 'red'
  ) +
  geom_line(
    aes(
      x = dataset$Level,
      y = predict(regressor, newdata = dataset)
    ),
    color = 'blue'
  ) +
  labs(
    title = 'Random Forest Regression',
    x = 'Level',
    y = 'Salary'
  ) +
  theme_minimal()

# ===========================
# 📊 High-Resolution Plot
# ===========================
X <- seq(min(dataset$Level), max(dataset$Level), 0.01)

ggplot() +
  geom_point(
    aes(x = dataset$Level, y = dataset$Salary),
    color = 'red'
  ) +
  geom_line(
    aes(
      x = X,
      y = predict(
        regressor,
        newdata = data.frame(Level = X)
      )
    ),
    color = 'blue'
  ) +
  labs(
    title = 'Random Forest Regression (High Resolution)',
    x = 'Level',
    y = 'Salary'
  ) +
  theme_minimal()
